In [1]:
import pandas as pd

data = pd.read_csv("thanksgiving.csv",encoding="Latin-1")
data[:3]

RespondentID Do you celebrate Thanksgiving?  \
0    4337954960                            Yes   
1    4337951949                            Yes   
2    4337935621                            Yes   

  What is typically the main dish at your Thanksgiving dinner?  \
0                                             Turkey             
1                                             Turkey             
2                                             Turkey             

  What is typically the main dish at your Thanksgiving dinner? - Other (please specify)  \
0                                                NaN                                      
1                                                NaN                                      
2                                                NaN                                      

  How is the main dish typically cooked?  \
0                                  Baked   
1                                  Baked   
2                                Roasted   

  How is the main dish typically cooked? - Other (please specify)  \
0                                                NaN                
1                                                NaN                
2                                                NaN                

  What kind of stuffing/dressing do you typically have?  \
0                                        Bread-based      
1                                        Bread-based      
2                                         Rice-based      

  What kind of stuffing/dressing do you typically have? - Other (please specify)  \
0                                                NaN                               
1                                                NaN                               
2                                                NaN                               

  What type of cranberry saucedo you typically have?  \
0                                               None   
1                             Other (please specify)   
2                                           Homemade   

  What type of cranberry saucedo you typically have? - Other (please specify)  \
0                                                NaN                            
1                    Homemade cranberry gelatin ring                            
2                                                NaN                            

          ...          \
0         ...           
1         ...           
2         ...           

  Have you ever tried to meet up with hometown friends on Thanksgiving night?  \
0                                                Yes                            
1                                                 No                            
2                                                Yes                            

  Have you ever attended a "Friendsgiving?"  \
0                                        No   
1                                        No   
2                                       Yes   

  Will you shop any Black Friday sales on Thanksgiving Day?  \
0                                                 No          
1                                                Yes          
2                                                Yes          

  Do you work in retail? Will you employer make you work on Black Friday?  \
0                     No                                              NaN   
1                     No                                              NaN   
2                     No                                              NaN   

  How would you describe where you live?      Age What is your gender?  \
0                               Suburban  18 - 29                 Male   
1                                  Rural  18 - 29               Female   
2                               Suburban  18 - 29                 Male   

  How much total combined money did all members of your HOUSEHOLD earn last year?  \
0                                 $

In [3]:
columns_index = pd.Series(data.columns)
print(columns_index)

0                                          RespondentID
1                        Do you celebrate Thanksgiving?
2     What is typically the main dish at your Thanks...
3     What is typically the main dish at your Thanks...
4                How is the main dish typically cooked?
5     How is the main dish typically cooked? - Other...
6     What kind of stuffing/dressing do you typicall...
7     What kind of stuffing/dressing do you typicall...
8     What type of cranberry saucedo you typically h...
9     What type of cranberry saucedo you typically h...
10                         Do you typically have gravy?
11    Which of these side dishes aretypically served...
12    Which of these side dishes aretypically served...
13    Which of these side dishes aretypically served...
14    Which of these side dishes aretypically served...
15    Which of these side dishes aretypically served...
16    Which of these side dishes aretypically served...
17    Which of these side dishes aretypically se

In [4]:
data["Do you celebrate Thanksgiving?"].value_counts()

Yes    980
No      78
Name: Do you celebrate Thanksgiving?, dtype: int64

In [6]:
data = data[data["Do you celebrate Thanksgiving?"]=="Yes"]

In [7]:
data["Do you celebrate Thanksgiving?"].value_counts()

Yes    980
Name: Do you celebrate Thanksgiving?, dtype: int64

In [8]:
data["What is typically the main dish at your Thanksgiving dinner?"].value_counts()

Turkey                    859
Other (please specify)     35
Ham/Pork                   29
Tofurkey                   20
Chicken                    12
Roast beef                 11
I don't know                5
Turducken                   3
Name: What is typically the main dish at your Thanksgiving dinner?, dtype: int64

In [9]:
tofurkey_eaters = data[data["What is typically the main dish at your Thanksgiving dinner?"]=="Tofurkey"]

In [10]:
tofurkey_eaters["Do you typically have gravy?"]

4      Yes
33     Yes
69      No
72      No
77     Yes
145    Yes
175    Yes
218     No
243    Yes
275     No
393    Yes
399    Yes
571    Yes
594    Yes
628     No
774     No
820     No
837    Yes
860     No
953    Yes
Name: Do you typically have gravy?, dtype: object

In [11]:
apple_isnull = data["Which type of pie is typically served at your Thanksgiving dinner? Please select all that apply. - Apple"].isnull()

In [12]:
pumpkin_isnull = data["Which type of pie is typically served at your Thanksgiving dinner? Please select all that apply. - Pumpkin"].isnull()

pecan_isnull = data["Which type of pie is typically served at your Thanksgiving dinner? Please select all that apply. - Pecan"].isnull()

In [25]:
ate_pies = apple_isnull & pumpkin_isnull & pecan_isnull

In [28]:
ate_pies.value_counts()

False    876
True     104
dtype: int64

In [75]:
def convert_age_range(age_value):
    if pd.isnull(age_value):
        return None
    first_char = age_value.split()[0]
    first_char = first_char.replace('+', '')
    return int(first_char)

In [76]:
int_data = data["Age"].apply(convert_age_range)

In [77]:
int_data.describe()

count    947.000000
mean      40.089757
std       15.352014
min       18.000000
25%       30.000000
50%       45.000000
75%       60.000000
max       60.000000
Name: Age, dtype: float64

## Age Breakdown

Above is a summary of the `int_data`, which only contains only the first number included in the original `Age` column in `data`. `Age` was recorded as age ranges. As such, the only values included in `int_data`, excluding null values, are `18` (`18-29`), `30` (`30-44`), `45` (`45-59`), and `60` (`60+`). So, the `pd.series.describe()` data above is not that helpful with regards to discerning Thanksgiving trends among certain age cohorts. 

It would be better to create 4 separate dataframes that had all rows where the `Age` column was equal to one age range. Then one can compare those for information on Thanksgiving dinner.

In [84]:
def convert_income_range(income_value):
    if pd.isnull(income_value) or income_value == 'Prefer not to answer':
        return None
    first_char = income_value.split()[0]
    first_char = first_char.replace('$', '').replace(',', '')
    return int(first_char)

In [85]:
int_income = data["How much total combined money did all members of your HOUSEHOLD earn last year?"].apply(convert_income_range)

In [86]:
data['int_data'] = int_data
data['int_income'] = int_income

In [87]:
data['int_income'].describe()

count       829.000000
mean      75965.018094
std       59068.636748
min           0.000000
25%       25000.000000
50%       75000.000000
75%      100000.000000
max      200000.000000
Name: int_income, dtype: float64

## Income breakdown

Converting and describing the income column in `data` was very similar ~identical~ to converting and describing the `Age` data. As such, `int_income`'s summary statistics suffer from the same problems that `int_data` does. Again, the data was originally collected as income ranges, but unlike `Age`, the income columns' ranges were not uniform. For example, the first explicit range, `0 to 9,999`, encompassed a step of 10,000. However, the second income group, `10,000 to 24,999`, was a step of 15,000. Every income group thereafter, with the exception of highest (`200,000 and up`), stepped up by 25,000. This might be why the `std` (standard deviation: `59068.636748`) is not equal to the range as it was for `int_data`.

I still don't think this is a true depiction of survey participants' incomes (or even the ranges they fall into). But it is better than the `Age` data. Based on the quartile information, we can see that most participants were grouped around the `75,000 to 99,000` range. Apply the standard deviation (which should encompass a little over 68% of survey participants if our data fits a bell curve), and most participants should fall within that range, +/- 60,000.


In [90]:
less_than_150k = data[data['int_income'] < 150000]

In [91]:
less_than_150k['How far will you travel for Thanksgiving?'].value_counts()

Thanksgiving is happening at my home--I won't travel at all                         281
Thanksgiving is local--it will take place in the town I live in                     203
Thanksgiving is out of town but not too far--it's a drive of a few hours or less    150
Thanksgiving is out of town and far away--I have to drive several hours or fly       55
Name: How far will you travel for Thanksgiving?, dtype: int64

In [92]:
more_than_150k = data[data['int_income'] > 150000]
more_than_150k['How far will you travel for Thanksgiving?'].value_counts()

Thanksgiving is happening at my home--I won't travel at all                         49
Thanksgiving is local--it will take place in the town I live in                     25
Thanksgiving is out of town but not too far--it's a drive of a few hours or less    16
Thanksgiving is out of town and far away--I have to drive several hours or fly      12
Name: How far will you travel for Thanksgiving?, dtype: int64

## Travel by Income Breakdown

DQ's assumption that people with less income would be more likely to travel farther for Thanksgiving was correct. Although I question their decision to split the incomes at 150k. I would have tried to split it between 5 figures and 6.

Still, those households with over 150k were split in about half between travelling and staying at home. For those households that make less than 150k, well over half travel for Thanksgiving.

In [2]:
import pandas as pd

data = pd.read_csv("thanksgiving.csv",encoding="Latin-1")

In [3]:
def convert_age_range(age_value):
    if pd.isnull(age_value):
        return None
    first_char = age_value.split()[0]
    first_char = first_char.replace('+', '')
    return int(first_char)

int_age = data["Age"].apply(convert_age_range)
data['int_age'] = int_age
friends_pt = data.pivot_table(index='Have you ever tried to meet up with hometown friends on Thanksgiving night?', columns='Have you ever attended a "Friendsgiving?"',values='int_age')

In [5]:
friends_pt

Have you ever attended a "Friendsgiving?"                  No        Yes
Have you ever tried to meet up with hometown fr...                      
No                                                  42.283702  37.010526
Yes                                                 41.475410  33.976744

In [6]:
def convert_income_range(income_value):
    if pd.isnull(income_value) or income_value == 'Prefer not to answer':
        return None
    first_char = income_value.split()[0]
    first_char = first_char.replace('$', '').replace(',', '')
    return int(first_char)

int_income = data["How much total combined money did all members of your HOUSEHOLD earn last year?"].apply(convert_income_range)
data['int_income'] = int_income
friends_pt_income = data.pivot_table(index='Have you ever tried to meet up with hometown friends on Thanksgiving night?', columns='Have you ever attended a "Friendsgiving?"',values='int_income')

In [7]:
friends_pt_income

Have you ever attended a "Friendsgiving?"                     No           Yes
Have you ever tried to meet up with hometown fr...                            
No                                                  78914.549654  72894.736842
Yes                                                 78750.000000  66019.736842

## Friendsgiving Breakdown

These questions are definitely aimed at the young. Millienials, probably. The lowest averages in age and income occur when a participant replied `Yes` to both questions `Have you ever tried to meet up with hometown friends on Thanksgiving night?` and `Have you ever attended a "Friendsgiving?"`. The inverse is also true, the highest income and age averages occured when participants replied `No` to both questions.

## Links of interest

- Original [dataset](https://github.com/fivethirtyeight/data/tree/master/thanksgiving-2015) from [FiveThirtyEight](http://fivethirtyeight.com/).

- Dataquest solutions [notebook](https://github.com/dataquestio/solutions/blob/master/Mission219Solution.ipynb).